### ISSUE TITLE
6.0.1 introduces false warnings

### ISSUE THREAD

**Issue by henrik242** [pmd/pmd#876]
After upgrading from 6.0.0 to 6.0.1 I get these warnings:
```
This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.0.1/pmd_userdocs_getting_started.html#incremental-analysis
Use Rule name category/java/errorprone.xml/AvoidBranchingStatementAsLastInLoop instead of the deprecated Rule name rulesets/java/basic.xml/AvoidBranchingStatementAsLastInLoop. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidBranchingStatementAsLastInLoop instead of the deprecated Rule name rulesets/java/basic.xml/AvoidBranchingStatementAsLastInLoop. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidDecimalLiteralsInBigDecimalConstructor instead of the deprecated Rule name rulesets/java/basic.xml/AvoidDecimalLiteralsInBigDecimalConstructor. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidDecimalLiteralsInBigDecimalConstructor instead of the deprecated Rule name rulesets/java/basic.xml/AvoidDecimalLiteralsInBigDecimalConstructor. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidMultipleUnaryOperators instead of the deprecated Rule name rulesets/java/basic.xml/AvoidMultipleUnaryOperators. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidMultipleUnaryOperators instead of the deprecated Rule name rulesets/java/basic.xml/AvoidMultipleUnaryOperators. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidUsingOctalValues instead of the deprecated Rule name rulesets/java/basic.xml/AvoidUsingOctalValues. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/AvoidUsingOctalValues instead of the deprecated Rule name rulesets/java/basic.xml/AvoidUsingOctalValues. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/BrokenNullCheck instead of the deprecated Rule name rulesets/java/basic.xml/BrokenNullCheck. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/BrokenNullCheck instead of the deprecated Rule name rulesets/java/basic.xml/BrokenNullCheck. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/CheckSkipResult instead of the deprecated Rule name rulesets/java/basic.xml/CheckSkipResult. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/CheckSkipResult instead of the deprecated Rule name rulesets/java/basic.xml/CheckSkipResult. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/ClassCastExceptionWithToArray instead of the deprecated Rule name rulesets/java/basic.xml/ClassCastExceptionWithToArray. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/ClassCastExceptionWithToArray instead of the deprecated Rule name rulesets/java/basic.xml/ClassCastExceptionWithToArray. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/DontUseFloatTypeForLoopIndices instead of the deprecated Rule name rulesets/java/basic.xml/DontUseFloatTypeForLoopIndices. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/DontUseFloatTypeForLoopIndices instead of the deprecated Rule name rulesets/java/basic.xml/DontUseFloatTypeForLoopIndices. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/JumbledIncrementer instead of the deprecated Rule name rulesets/java/basic.xml/JumbledIncrementer. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/JumbledIncrementer instead of the deprecated Rule name rulesets/java/basic.xml/JumbledIncrementer. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/MisplacedNullCheck instead of the deprecated Rule name rulesets/java/basic.xml/MisplacedNullCheck. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/MisplacedNullCheck instead of the deprecated Rule name rulesets/java/basic.xml/MisplacedNullCheck. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/OverrideBothEqualsAndHashcode instead of the deprecated Rule name rulesets/java/basic.xml/OverrideBothEqualsAndHashcode. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/OverrideBothEqualsAndHashcode instead of the deprecated Rule name rulesets/java/basic.xml/OverrideBothEqualsAndHashcode. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/ReturnFromFinallyBlock instead of the deprecated Rule name rulesets/java/basic.xml/ReturnFromFinallyBlock. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/ReturnFromFinallyBlock instead of the deprecated Rule name rulesets/java/basic.xml/ReturnFromFinallyBlock. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/UnconditionalIfStatement instead of the deprecated Rule name rulesets/java/basic.xml/UnconditionalIfStatement. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/errorprone.xml/UnconditionalIfStatement instead of the deprecated Rule name rulesets/java/basic.xml/UnconditionalIfStatement. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/AvoidThreadGroup instead of the deprecated Rule name rulesets/java/basic.xml/AvoidThreadGroup. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/AvoidThreadGroup instead of the deprecated Rule name rulesets/java/basic.xml/AvoidThreadGroup. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/DontCallThreadRun instead of the deprecated Rule name rulesets/java/basic.xml/DontCallThreadRun. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/DontCallThreadRun instead of the deprecated Rule name rulesets/java/basic.xml/DontCallThreadRun. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/DoubleCheckedLocking instead of the deprecated Rule name rulesets/java/basic.xml/DoubleCheckedLocking. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/multithreading.xml/DoubleCheckedLocking instead of the deprecated Rule name rulesets/java/basic.xml/DoubleCheckedLocking. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/bestpractices.xml/AvoidUsingHardCodedIP instead of the deprecated Rule name rulesets/java/basic.xml/AvoidUsingHardCodedIP. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/bestpractices.xml/AvoidUsingHardCodedIP instead of the deprecated Rule name rulesets/java/basic.xml/AvoidUsingHardCodedIP. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/bestpractices.xml/CheckResultSet instead of the deprecated Rule name rulesets/java/basic.xml/CheckResultSet. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/bestpractices.xml/CheckResultSet instead of the deprecated Rule name rulesets/java/basic.xml/CheckResultSet. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/codestyle.xml/ExtendsObject instead of the deprecated Rule name rulesets/java/basic.xml/ExtendsObject. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/codestyle.xml/ExtendsObject instead of the deprecated Rule name rulesets/java/basic.xml/ExtendsObject. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/codestyle.xml/ForLoopShouldBeWhileLoop instead of the deprecated Rule name rulesets/java/basic.xml/ForLoopShouldBeWhileLoop. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/codestyle.xml/ForLoopShouldBeWhileLoop instead of the deprecated Rule name rulesets/java/basic.xml/ForLoopShouldBeWhileLoop. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/performance.xml/BigIntegerInstantiation instead of the deprecated Rule name rulesets/java/basic.xml/BigIntegerInstantiation. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/performance.xml/BigIntegerInstantiation instead of the deprecated Rule name rulesets/java/basic.xml/BigIntegerInstantiation. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/performance.xml/BooleanInstantiation instead of the deprecated Rule name rulesets/java/basic.xml/BooleanInstantiation. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/performance.xml/BooleanInstantiation instead of the deprecated Rule name rulesets/java/basic.xml/BooleanInstantiation. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/design.xml/CollapsibleIfStatements instead of the deprecated Rule name rulesets/java/basic.xml/CollapsibleIfStatements. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/design.xml/CollapsibleIfStatements instead of the deprecated Rule name rulesets/java/basic.xml/CollapsibleIfStatements. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/design.xml/SimplifiedTernary instead of the deprecated Rule name rulesets/java/basic.xml/SimplifiedTernary. PMD 7.0.0 will remove support for this deprecated Rule name usage.
Use Rule name category/java/design.xml/SimplifiedTernary instead of the deprecated Rule name rulesets/java/basic.xml/SimplifiedTernary. PMD 7.0.0 will remove support for this deprecated Rule name usage.
```
None of these deprecated rules are in my pmd-ruleset.xml.  

But some of the correct rules are there, e.g.
```
  <rule ref="category/java/errorprone.xml/BrokenNullCheck">
    <priority>1</priority>
  </rule>
```

Any ideas why this happens?

**Comment by jsotuyod** [comment]
@henrik242 Please, include your complete current ruleset.

**Comment by henrik242** [comment]
@jsotuyod  Sure, here it is. (It worked perfectly in 6.0.0.)

[pmd-ruleset.xml.txt](https://github.com/pmd/pmd/files/1655794/pmd-ruleset.xml.txt)

**Comment by adangel** [comment]
@henrik242 thanks for the ruleset. However, I can't reproduce it when running PMD from commandline:

```
$ pmd-bin-6.0.1/bin/run.sh pmd -d src -f text -R pmd-ruleset.xml 
Jan 23, 2018 2:19:30 PM net.sourceforge.pmd.cache.NoopAnalysisCache <init>
WARNUNG: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/pmd-6.0.1/pmd_userdocs_getting_started.html#incremental-analysis
```

How are you executing PMD? Ant? Maven? Gradle?
If using Gradle, make sure to not accidentally use the default rulesets additionally: The [PmdExtension](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.PmdExtension.html) has two ways of configuring the rulesets: `ruleSetFiles` and `ruleSets`. If you configure `ruleSetFiles = "pmd-ruleset.xml"` then you'll need to clear the other property: `ruleSets = []`, otherwise all rulesets will be executed (yours and additionally basic and braces).

**Comment by henrik242** [comment]
That solved it!  I had this config:
```
tasks.withType(Pmd) {
    it.ruleSetFiles = files("${myProjectRoot}/shared/config/pmd-ruleset.xml")
}
```
By adding `it.ruleSets = []` the warnings went away.  Why did this change in 6.0.1, by the way?  I've been using PMD for ages, but never had to set `ruleSets`.

**Comment by adangel** [comment]
In 6.0.1, we added warnings for each single rule contained in a deprecated ruleset. Before we would have only printed a warning for (deprecated) single rule references.

Glad that it works now!

**Comment by adangel** [comment]
Related issues/fixes: #865 and #805

**Comment by jsotuyod** [comment]
@henrik242 notice that we only added the deprecation warnings for ease of migration. You have been using rules outside your defined ruleset for ages due to Gradle's implementation.

**Comment by henrik242** [comment]
@jsotuyod Hm! Just a thought: Why is PMD pointing to outdated ruleSets when I don't set `ruleSets = []`?

**Comment by jsotuyod** [comment]
@henrik242 that's Gradle

![image](https://user-images.githubusercontent.com/802626/35284097-a9551d58-0038-11e8-8ac6-2e213d07d85c.png)

https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.PmdExtension.html#org.gradle.api.plugins.quality.PmdExtension:ruleSets

If you don't change that value, the default remains; and Gradle itself merges all rulesets / rulesetfiles  / rulesetconfig and passes them all to PMD.

**Comment by henrik242** [comment]
Ah, OK.  So I'll just have to wait for Gradle to try the PMD 6.0.x upgrade again, ref. https://github.com/gradle/gradle/pull/3125, then I can remove the `ruleSets = []` line.  Thanks for clearing that up :)

**Comment by LeshaRB** [comment]
I add

```
pmd {
  ignoreFailures = true
  toolVersion = "6.0.1"
  sourceSets = [sourceSets.main, sourceSets.test]
  reportsDir = file("$project.buildDir/pmdReports")
  ruleSetFiles = files("$configDir/pmd-ruleset.xml")
}

tasks.withType(Pmd) {
  reports {
    xml.enabled = false
    html.enabled = true
  }
}
```

But it don't work for me

**Comment by jsotuyod** [comment]
@LeshaRB you are not including `ruleSets = []` in the  `pmd { … }` block.

**Comment by LeshaRB** [comment]
`ruleSets = [] `
or 
`ruleSets = [“basic”, “braces”] `
or
`ruleSets = [“java-basic”, “-javabraces”]`

**Comment by jsotuyod** [comment]
@LeshaRB I'm having trouble understanding your last comment...

If your config is:

```
pmd {
  ignoreFailures = true
  toolVersion = "6.0.1"
  sourceSets = [sourceSets.main, sourceSets.test]
  reportsDir = file("$project.buildDir/pmdReports")
  ruleSetFiles = files("$configDir/pmd-ruleset.xml")
  ruleSets = []
}

tasks.withType(Pmd) {
  reports {
    xml.enabled = false
    html.enabled = true
  }
}
```

**AND** your `pmd-ruleset.xml` config is up to date for PMD 6, you should see no warnings.

If you do, please provide the contents of your `pmd-ruleset.xml` file

**Comment by LeshaRB** [comment]
@jsotuyod Ok thanks

Sorry for my English
In some place for example https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.PmdExtension.html
use  `ruleSets = ["basic", "braces"]`

So my question was

What is the right use?

ruleSets = []
or
ruleSets = [“basic”, “braces”]
or
ruleSets = [“java-basic”, “java-braces”]

**Comment by LeshaRB** [comment]
Where I can take last pmd-ruleset.xml?

**Comment by jsotuyod** [comment]
@LeshaRB that file is yours.... it's you who wrote on your `build.gradle` the line `ruleSetFiles = files("$configDir/pmd-ruleset.xml")`

As for `ruleSets`, the correct is to reset it to empty `ruleSets = []`. Otherwise, the default will apply the deprecated java-basic ruleset, producing the deprecation warnings seen on the first post of this thread.

If you want those rules, you should include them in your ruleset XML manually using their new location.
